datadog/fwprovider/resource_datadog_on_call_schedule.go

```diff
@@ -0,0 +1,803 @@
+		}
+		state.Layers[i] = newLayerModel(layersByID[layer.GetId()], membersByID, layerExistingEffectiveDate)
+	}
+	return state
+}
+
+func newLayerModel(layer *datadogV2.Layer, membersByID map[string]*datadogV2.ScheduleMember, layerExistingEffectiveDate customtypes.BackwardRFC3339Date) *layersModel {
+	membersData := layer.GetRelationships().Members.GetData()
+	members := make([]*membersModel, len(membersData))
+	for j, member := range membersData {
+		includedMember := membersByID[member.GetId()]
+		var user *userModel
+		if includedMember.GetRelationships().User != nil {
+			user = &userModel{Id: types.StringValue(*includedMember.GetRelationships().User.GetData().Id)}
+		}
+		members[j] = &membersModel{User: user}
+	}
+	restrictions := layer.GetAttributes().Restrictions
+	restrictionsModels := make([]*restrictionsModel, len(restrictions))
+	for j, restriction := range restrictions {
+		restrictionsModels[j] = &restrictionsModel{
+			EndDay:    types.StringValue(string(restriction.GetEndDay())),
+			EndTime:   types.StringValue(restriction.GetEndTime()),
+			StartDay:  types.StringValue(string(restriction.GetStartDay())),
+			StartTime: types.StringValue(restriction.GetStartTime()),
+		}
+	}
+	interval := layer.GetAttributes().Interval
+
+	var endDateStringValue types.String
+	endDate := layer.GetAttributes().EndDate
+	if endDate != nil && !endDate.IsZero() {
+		endDateStringValue = types.StringValue(formatTime(*endDate))
+	} else {
+		endDateStringValue = types.StringNull()
+	}
+
+	appliedEffectiveDate := customtypes.NewBackwardRFC3339Date(formatTime(layer.Attributes.GetEffectiveDate()))
+
+	// If the effective date is not set, use the applied effective date
+	// Otherwise, use the existing effective date
+	// The effective date is irrelevant to the state (applied_effective_date is), we just need to be sure to make it
+	// consistent for Terraform.
+	var effectiveDate customtypes.BackwardRFC3339Date
+	if layerExistingEffectiveDate.ValueString() != "" {
+		effectiveDate = customtypes.NewBackwardRFC3339Date(layerExistingEffectiveDate.ValueString())
+	} else {
+		effectiveDate = appliedEffectiveDate
+	}
+
+	return &layersModel{
+		Id:                   types.StringValue(layer.GetId()),
+		AppliedEffectiveDate: appliedEffectiveDate,
+		EffectiveDate:        effectiveDate,
+		EndDate:              endDateStringValue,
+		Name:                 types.StringValue(layer.Attributes.GetName()),
+		RotationStart:        types.StringValue(formatTime(layer.Attributes.GetRotationStart())),
+		Members:              members,
+		Restrictions:         restrictionsModels,
+		Interval:             &intervalModel{Days: types.Int32Value(int32(interval.GetDays())), Seconds: types.Int64Value(interval.GetSeconds())},
+	}
+}
+
+func formatTime(t time.Time) string {
+	return t.Format(time.RFC3339)
+}
+
+func (r *onCallScheduleResource) buildOnCallScheduleRequestBody(ctx context.Context, state *onCallScheduleModel) (*datadogV2.ScheduleCreateRequest, diag.Diagnostics) {
```

Comment:
a lot of code is duplicated between Create and Update because the golang sdk use different go types for both actions. This might be simplified by potentially sharing some types in between the 2 operations. (Types are not the same as some minor differences exsit between the 2 operations, for example an PUT/Update might include a layer.id field; a POST/Create can't.